### PR TITLE
[Task]15209420: add web snippet prefix and config

### PR DESCRIPTION
**Pull request by Karlie-777**
1. change config name `snippet` to `webInstrumentation`
2. old config/methods should still work
3. add client config for JS SDK
4. add client config for snippet src
5. add more test coverages

### CODE REVIEW

File: Declarations/Interfaces.ts [hunk 1]
```diff
@@ -135,14 +135,41 @@ export interface IBaseConfig {
...
-     * Application Insights resource connection string for web snippet
+     * @deprecated, Please use webInstrumentationConnectionString instead
+     * Application Insights resource connection string for web snippet, this config is NOT exposed in documentation after version 2.3.5
      * Note: if no valid connection string is provided here, web snippet will use the connection string during initializing Nodejs SDK
      */
     webSnippetConnectionString?: string;
+    /**
+     * Enable web instrumentation and automatic monitoring, default to false
+     */
+    enableWebInstrumentation: boolean;
```
Comment (Karlie-777) [inline Declarations/Interfaces.ts:151]: change the config name, the old config name still works (marked as deprecated)

File: AutoCollection/WebSnippet.ts [hunk 2]
```diff
@@ -68,13 +69,98 @@ class WebSnippet {
...
+        let rpStr = this._getResourceProvider();
+        let snippetReplacedStr = `${this._webInstrumentationIkey}\",\r\n${configStr} disableIkeyDeprecationMessage: true,\r\n sdkExtension: \"${rpStr}${osStr}r_n_`;
+        let replacedSnippet = webSnippet.replace("INSTRUMENTATION_KEY", snippetReplacedStr);
+        if (this._clientWebInstrumentationSrc) {
+            return replacedSnippet.replace(`${Constants.WEB_INSTRUMENTATION_DEFAULT_SOURCE}.2.min.js`,this._clientWebInstrumentationSrc);
+        }
+        return replacedSnippet;
+    }
+
+    private _getOsString() {
```
Comment (hectorhdzg) [inline AutoCollection/WebSnippet.ts:101]: We use this line to determine if Linux/Windows in Agent

  const isWindows = process.platform === 'win32';

File: AutoCollection/WebSnippet.ts [hunk 3]
```diff
@@ -68,13 +69,98 @@ class WebSnippet {
...
+            // if has any errors here, web Instrumentation will be disabled.
+            this._isEnabled = false;
+            Logging.info("Parse client web instrumentation error. Web Instrumentation is disabled");
+        }
+        return configStr;
+    }
+
+    private _getResourceProvider() {
+        let resourceProvider = "u";
+        if (process.env.WEBSITE_SITE_NAME) { // Web apps
```
Comment (hectorhdzg) [inline AutoCollection/WebSnippet.ts:158]: WEBSITE_SITE_NAME should exist in both Web Apps and Functions, is not the case anymore?

File: Declarations/Constants.ts [hunk 4]
```diff
@@ -199,4 +199,7 @@ export const MessageBusDestination = "message_bus.destination";
   * AI time since enqueued attribute.
   * @internal
   */
- export const TIME_SINCE_ENQUEUED = "timeSinceEnqueued";
\ No newline at end of file
+ export const TIME_SINCE_ENQUEUED = "timeSinceEnqueued";
+
+ export const WEB_INSTRUMENTATION_DEFAULT_SOURCE= "https://js.monitor.azure.com/scripts/b/ai";
+ export const WEB_INSTRUMENTATION_DEPRECATE_SOURCE= "https://az416426.vo.msecnd.net/scripts/b/ai";
```
Comment (hectorhdzg) [inline Declarations/Constants.ts:205]:
```suggestion
 export const WEB_INSTRUMENTATION_DEPRECATED_SOURCE= "https://az416426.vo.msecnd.net/scripts/b/ai";
```

File: Declarations/Interfaces.ts [hunk 5]
```diff
@@ -135,14 +135,53 @@ export interface IBaseConfig {
...
+     /**
+     * Application Insights web Instrumentation CDN url
+     * NOTE: this config can be changed from env variable: APPLICATIONINSIGHTS_WEB_INSTRUMENTATION_SOURCE or Json Config: webInstrumentationSrc
+     * If no resouce is provided here, default CDN endpoint: https://js.monitor.azure.com/scripts/b/ai will be used
+     * see more details at: https://github.com/microsoft/ApplicationInsights-JS
+     */
+    webInstrumentationSrc?: string;
+}
+
+export interface IwebInstrumentationConfig {
```
Comment (hectorhdzg) [inline Declarations/Interfaces.ts:175]:
```suggestion
export interface IWebInstrumentationConfig {
```

File: Declarations/Interfaces.ts [hunk 6]
```diff
@@ -135,14 +135,53 @@ export interface IBaseConfig {
...
+    /**
+     * Name of Application Insights web Instrumentation config to be changed
+     * see more Application Insights web Instrumentation config details at: https://github.com/microsoft/ApplicationInsights-JS#configuration
+     */
+    configName: string;
```
Comment (hectorhdzg) [inline Declarations/Interfaces.ts:180]:
```suggestion
    name: string;
```

File: Library/Config.ts [hunk 7]
```diff
@@ -56,6 +56,10 @@ class Config implements IConfig {
     public disableStatsbeat: boolean;
     public extendedMetricDisablers: string;
     public quickPulseHost: string;
+    public enableWebInstrumentation: boolean;
```
Comment (hectorhdzg) [inline Library/Config.ts:59]: Add new configs in README file

**Review by hectorhdzg — APPROVED** [review]
LGTM

**Pull request merged**